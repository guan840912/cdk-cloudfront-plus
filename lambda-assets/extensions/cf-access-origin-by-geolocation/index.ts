export async function handler(event: any) {
  const request = event.Records[0].cf.request;
  const countryCodeTable = process.env.COUNTRY_CODE_TABLE
  let newDomainName: string;
  let viewerCountry: string;

  if (request.headers['cloudfront-viewer-country']) {
    viewerCountry = request.headers['cloudfront-viewer-country'][0].value;
    console.log('Got viewer country: %j', viewerCountry )
    newDomainName = countryCodeTable[viewerCountry]
    console.log('new domain name: %j', newDomainName)
    if (newDomainName) {
      request.origin.custom.domainName = newDomainName;
      request.headers['host'] = [{ key: 'host', value: newDomainName }];
    }
  }

  return request
};
