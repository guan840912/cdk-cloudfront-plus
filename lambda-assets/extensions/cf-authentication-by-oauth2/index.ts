/*
 *  Reference: https://github.com/experoinc/aws-lambda-edge-oauth
 */
const querystring = require("querystring");
const https = require("https");
const jsonwebtoken = require("jsonwebtoken");
const escape = require("lodash.escape");

const PUBLIC_PATHS = [/\/favicons\//];

function redirect(newLocation, cookies) {
  const result = {
    status: "302",
    statusDescription: "Found",
    headers: {
      location: [
        {
          key: "Location",
          value: newLocation,
        },
      ],
    },
  };

  if (cookies) {
    result.headers["set-cookie"] = cookies.map((c) => ({
      key: "set-cookie",
      value: `${c.name}=${c.value}`,
    }));
  }

  return result;
}

function respond(status, statusText, title, message) {
  const body = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>${escape(title)}</title>
  </head>
  <body>
    <p>${escape(message)}</p>
  </body>
</html>`;

  const result = {
    status: status,
    statusText: statusText,
    headers: {
      "cache-control": [
        {
          key: "Cache-Control",
          value: "max-age=100",
        },
      ],
      "content-type": [
        {
          key: "Content-Type",
          value: "text/html",
        },
      ],
    },
    body: body,
  };

  return result;
}

function parseCookies(headers) {
  const parsedCookie = {};
  if (headers.cookie) {
    headers.cookie[0].value.split(";").forEach((cookie) => {
      if (cookie) {
        const parts = cookie.split("=");
        parsedCookie[parts[0].trim()] = parts[1].trim();
      }
    });
  }

  return parsedCookie;
}

function validateToken(config, token) {
  if (config.DEBUG_ENABLE) console.log("validateToken: enter, token = " + token + ", config.JWT_ARGORITHM = " + config.JWT_ARGORITHM);

  if (config.DEBUG_ENABLE) {
    const decodedjwt = jsonwebtoken.decode(token, {complete: true});
    console.log("validateToken: token header = (next line)");
    console.log(decodedjwt.header);
  }
  try {
    const decoded = jsonwebtoken.verify(token, config.CLIENT_PUBLIC_KEY, {
      algorithms: [config.JWT_ARGORITHM],
    });

    if (config.DEBUG_ENABLE) console.log("validateToken: return true");
    return true;
  } catch (err) {
    if (config.DEBUG_ENABLE) console.log("validateToken: return false, err = " + err);
    return false;
  }
}

function validateCookie(config, cookie) {
  return !!cookie && validateToken(config, cookie);
}

function loginCallback(config, request, callback) {
  if (config.DEBUG_ENABLE) console.log("loginCallback: enter");

  if (request.uri !== config.CALLBACK_PATH) {
    console.log("loginCallback: is not callback path");
    return false;
  } // unhandled

  if (config.DEBUG_ENABLE) console.log("loginCallback: request.querystring = " + request.querystring);

  let params;
  try {
    params = querystring.parse(request.querystring);
    if (params.error) {
      if (config.DEBUG_ENABLE) console.log("loginCallback: params.error");
      callback(
        null,
        respond(401, "Unauthorized", params.error, params.error_description)
      );
      return true; // handled
    }
    if (!params.code) {
      if (config.DEBUG_ENABLE) console.log("loginCallback: params.code is missing!!");
      return false; // unhandled
    }
  } catch (err) {
    return false; // unhandled
  }

  // Call identity provider to get JWT token
  const headers = request.headers;
  const postData = querystring.stringify({
    grant_type: "authorization_code",
    code: params.code,
    client_id: config.CLIENT_ID,
    client_secret: config.CLIENT_SECRET,
    redirect_uri: `https://${headers.host[0].value}${config.CALLBACK_PATH}`,
  });
  const postOptions = {
    host: config.CLIENT_DOMAIN,
    port: 443,
    path: "/oauth/token",
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "Content-Length": postData.length,
    },
  };
  const req = https.request(postOptions, (res) => {
    if (config.DEBUG_ENABLE) console.log("loginCallback: res.statusCode = " + res.statusCode);
    if (res.statusCode >= 300) {
      return callback(
        null,
        respond(
          500,
          "Internal Server Error",
          "Bad token response",
          `Bad Token Response: ${res.statusCode} ${res.statusText}`
        )
      );
    }

    let body = "";
    res.on("data", (d) => (body += d));
    res.on("end", () => {
      try {
        const json = JSON.parse(body);
        if (config.DEBUG_ENABLE) console.log("loginCallback: parse body = (next line)");
        if (config.DEBUG_ENABLE) console.log(json);
        const token = json.id_token;
        if (config.DEBUG_ENABLE) console.log("loginCallback: token = " + token);

        if (!token) {
          if (config.DEBUG_ENABLE) console.log("loginCallback: no token!");
          return callback(
            null,
            respond(401, "Unauthorized", "Unauthorized", body)
          );
        }

        // store this in a cookie, then redirect the user
        const dest = `https://${headers.host[0].value}${params.dest || "/"}`;
        if (config.DEBUG_ENABLE) console.log("loginCallback: redirect to dest = " + dest + " with session token");

        callback(
          null,
          redirect(dest, [{ name: "session-token", value: token }])
        );
      } catch (e) {
        callback(e, null);
      }
    });
  });
  req.on("error", (e) => callback(e, null));
  req.write(postData);
  req.end();

  if (config.DEBUG_ENABLE) console.log("loginCallback: finish with true.");

  return true; // handled
}

function redirectIfNotAuthenticated(config, request, callback) {
  if (config.DEBUG_ENABLE) console.log("redirectIfNotAuthenticated: enter.");

  const headers = request.headers;

  /* Check for session-id in request cookie in viewer-request event,
   * if session-id is absent, redirect the user to sign in page with original
   * request sent as redirect_url in query params.
   */

  /* Check for session-id in cookie, if present then proceed with request */
  const parsedCookies = parseCookies(headers);
  if (config.DEBUG_ENABLE) console.log("redirectIfNotAuthenticated: parsedCookies = (next line)");
  if (config.DEBUG_ENABLE) console.log(parsedCookies);
  if (validateCookie(config, parsedCookies && parsedCookies["session-token"])) {
    if (config.DEBUG_ENABLE) console.log("redirectIfNotAuthenticated: validateCookie: false");
    return false; // not handled
  }

  // User is not authenticated.
  /* URI encode the original request so we can send as query param for when user is finally logged in */
  const encodedRedirectUrl = encodeURIComponent(
    request.querystring ? `${request.uri}?${request.querystring}` : request.uri
  );
  const callbackUrl = `https://${headers.host[0].value}${config.CALLBACK_PATH}?dest=${encodedRedirectUrl}`;
  const encodedCallback = encodeURIComponent(callbackUrl);
  const redirectUrl = `${config.AUTHORIZE_URL}${config.AUTHORIZE_PARAMS}&client_id=${config.CLIENT_ID}&redirect_uri=${encodedCallback}`;

  if (config.DEBUG_ENABLE) console.log("redirectIfNotAuthenticated: redirect to AUTHORIZE_URL with empty session token. redirectUrl = " + redirectUrl);
  callback(null, redirect(redirectUrl, [{ name: "session-token", value: "" }]));

  return true; // handled
}

function allowPublicPaths(config, request, callback) {
  if (config.DEBUG_ENABLE) console.log("allowPublicPaths: enter.");
  if (PUBLIC_PATHS.find(pattern => pattern.test(request.uri))) {
    if (config.DEBUG_ENABLE) console.log("allowPublicPaths: callback request and return true.");
    callback(null, request);
    return true;
  }else{
    if (config.DEBUG_ENABLE) console.log("allowPublicPaths: return false.");
    return false;
  }
}

export function handler(event: any, context: any, callback: any) {
  const config = {
    CLIENT_ID: process.env.CLIENT_ID,
    CLIENT_SECRET: process.env.CLIENT_SECRET,
    CLIENT_DOMAIN: process.env.CLIENT_DOMAIN,
    CLIENT_PUBLIC_KEY: new Buffer(process.env.CLIENT_PUBLIC_KEY, 'base64').toString(),
    CALLBACK_PATH: process.env.CALLBACK_PATH,
    JWT_ARGORITHM: process.env.JWT_ARGORITHM,
    AUTHORIZE_URL: process.env.AUTHORIZE_URL,
    AUTHORIZE_PARAMS: new Buffer(process.env.AUTHORIZE_PARAMS, 'base64').toString(),
    DEBUG_ENABLE: process.env.DEBUG_ENABLE,
  };

  const request = event.Records[0].cf.request;

  if (config.DEBUG_ENABLE) console.log("Event: %j", event);
  if (config.DEBUG_ENABLE) console.log("Context: %j", context);
  if (config.DEBUG_ENABLE) console.log("request URI = ", request.uri);

  if (
    !allowPublicPaths(config, request, callback) &&
    !loginCallback(config, request, callback) &&
    !redirectIfNotAuthenticated(config, request, callback)
  ) {
    if (config.DEBUG_ENABLE) console.log("hanlder: callback request.");
    callback(null, request);
  }
}
