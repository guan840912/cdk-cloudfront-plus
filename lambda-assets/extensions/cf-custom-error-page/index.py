#!/usr/bin/python
# -*- coding: utf-8 -*-
import json


# source from https://aws.amazon.com/blogs/networking-and-content-delivery/customize-403-error-pages-from-amazon-cloudfront-origin-with-lambdaedge/
def handler(event, context):
    
    print(event)

    response = event['Records'][0]['cf']['response']
    request = event['Records'][0]['cf']['request']

    '''
    This function updates your HTTP status code in the response to a 404, redirecting it to another path (cache behavior) that has a different origin configured. Note the following:
    1. The function is triggered by an origin response
    2. The response status from the origin server is an error status code (4xx)
    '''

    if int(response['status']) >= 400 and int(response['status']) <= 499:
    
        redirect_path = '/404.html'

        response['status'] = 302
        response['statusDescription'] = 'Found'

        # Drop the body as it is not required for redirects
        response['body'] = ''
        response['headers']['location'] = [{'key': 'Location', 'value': redirect_path}]

    print(response)
    return response