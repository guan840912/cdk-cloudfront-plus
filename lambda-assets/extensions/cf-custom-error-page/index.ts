export async function handler(event: any) {
  // Extract the request from the CloudFront event that is sent to Lambda@Edge

  console.log('Event', event);

  var response = event.Records[0].cf.response;
  /**
    This function updates your HTTP status code in the response to a 404, redirecting it to another path (cache behavior) that has a different origin configured. Note the following:
    1. The function is triggered by an origin response
    2. The response status from the origin server is an error status code (4xx)
   */
  if (Number(response.status) >= 400 && Number(response.status) <= 499) {
    const redirectPath = '/404.html';
    response.status = 302;
    response.statusDescription = 'Found';

    //Drop the body as it is not required for redirects
    response.body = '';
    response.headers.location = [{ key: 'Location', value: redirectPath }];
  };

  console.log(response);
  return response;
}