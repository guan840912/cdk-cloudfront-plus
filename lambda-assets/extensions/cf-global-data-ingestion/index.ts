import * as AWS from 'aws-sdk';

const kinesisFirehose = new AWS.Firehose({region: 'us-east-1'});
const deliveryStreamName = process.env.DELIVERY_STREAM_NAME;

export async function handler(event: any) {
  const request = event.Records[0].cf.request;
  const body = Buffer.from(request.body.data, 'base64').toString('utf-8');

  const params = {
    DeliveryStreamName: deliveryStreamName,
    Record: {
      Data: JSON.stringify({ body }) + '\n',
    },
  };

  await kinesisFirehose.putRecord(params).promise()

  return {
    body: '',
    bodyEncoding: 'text',
    status: '200',
  };
}