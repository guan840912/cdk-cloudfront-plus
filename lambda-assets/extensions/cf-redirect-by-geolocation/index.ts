export async function handler(event: any) {
  const request = event.Records[0].cf.request;
  const countryCodeTable = process.env.COUNTRY_CODE_TABLE;
  let newDomainName: string;
  let viewerCountry: string;
  let response: { [key: string]: any } = {}
  if (request.headers['cloudfront-viewer-country']) {
    viewerCountry = request.headers['cloudfront-viewer-country'][0].value;
    console.log('Got viewer country: %j', viewerCountry)
    newDomainName = countryCodeTable[viewerCountry]
    console.log('new domain name: %j', newDomainName)
    if (newDomainName) {
      response = {
        status: '302',
        statusDescription: 'Found',
        headers: {
            location: [{
                key: 'Location',
                value: newDomainName,
            }],
        },
      };
    }
  }
  return response;
}