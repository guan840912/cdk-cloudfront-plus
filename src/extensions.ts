import * as path from 'path';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as lambda from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as cdk from '@aws-cdk/core';
import { ServerlessApp } from './';

/**
 * The directory for all extensions lambda assets
 */
const EXTENSION_ASSETS_PATH = path.join(__dirname, '../lambda-assets/extensions');

/**
 * The Extension interface
 */
export interface IExtensions {
  /**
   * Lambda function ARN for this extension
   */
  readonly functionArn: string;
  /**
   * Lambda function version for the function
   */
  readonly functionVersion: lambda.Version;
  /**
   * The Lambda edge event type for this extension
   */
  readonly eventType: cf.LambdaEdgeEventType;
};

/**
 * The modify response header extension
 * @see https://github.com/awslabs/aws-cloudfront-extensions/tree/main/edge/nodejs/modify-response-header
 * @see https://console.aws.amazon.com/lambda/home#/create/app?applicationId=arn:aws:serverlessrepo:us-east-1:418289889111:applications/modify-response-header
 */
export class ModifyResponseHeader extends ServerlessApp implements IExtensions {
  readonly functionArn: string;
  readonly functionVersion: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id, {
      applicationId: 'arn:aws:serverlessrepo:us-east-1:418289889111:applications/modify-response-header',
      semanticVersion: '1.0.0',
    });
    const stack = cdk.Stack.of(scope);
    this.functionArn = this.resource.getAtt('Outputs.ModifyResponseHeaderFunctionARN').toString();
    this.functionVersion = bumpFunctionVersion(stack, id, this.functionArn);
    this.eventType = cf.LambdaEdgeEventType.ORIGIN_RESPONSE;
  }
}

/**
 * Construct properties for AntiHotlinking
 */
export interface AntiHotlinkingProps {
  /**
   * Referer allow list with wildcard(* and ?) support i.e. `example.com` or `exa?ple.*`
   */
  readonly referer: string[];
}

/**
 * The Anti-Hotlinking extension
 * @see https://github.com/awslabs/aws-cloudfront-extensions/tree/main/edge/nodejs/anti-hotlinking
 * @see https://console.aws.amazon.com/lambda/home#/create/app?applicationId=arn:aws:serverlessrepo:us-east-1:418289889111:applications/anti-hotlinking
 */
export class AntiHotlinking extends ServerlessApp implements IExtensions {
  readonly functionArn: string;
  readonly functionVersion: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  constructor(scope: cdk.Construct, id: string, props: AntiHotlinkingProps) {
    super(scope, id, {
      applicationId: 'arn:aws:serverlessrepo:us-east-1:418289889111:applications/anti-hotlinking',
      semanticVersion: '1.2.5',
      parameters: {
        RefererList: props.referer.join(','),
      },
    });
    const stack = cdk.Stack.of(scope);
    this.functionArn = this.resource.getAtt('Outputs.AntiHotlinking').toString();
    this.functionVersion = bumpFunctionVersion(stack, id, this.functionArn);
    this.eventType = cf.LambdaEdgeEventType.VIEWER_REQUEST;
  }
}

/**
 * Security Headers extension
 * @see https://console.aws.amazon.com/lambda/home?region=us-east-1#/create/app?applicationId=arn:aws:serverlessrepo:us-east-1:418289889111:applications/add-security-headers
 * @see https://aws.amazon.com/tw/blogs/networking-and-content-delivery/adding-http-security-headers-using-lambdaedge-and-amazon-cloudfront/
 */
export class SecurtyHeaders extends ServerlessApp implements IExtensions {
  readonly functionArn: string;
  readonly functionVersion: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id, {
      applicationId: 'arn:aws:serverlessrepo:us-east-1:418289889111:applications/add-security-headers',
      semanticVersion: '1.0.0',
    });
    const stack = cdk.Stack.of(scope);
    this.functionArn = this.resource.getAtt('Outputs.AddSecurityHeaderFunction').toString();
    this.functionVersion = bumpFunctionVersion(stack, id, this.functionArn);
    this.eventType = cf.LambdaEdgeEventType.ORIGIN_RESPONSE;
  }
}

export interface CustomProps {
  /**
   * Specify your Lambda function.
   *
   * You can specify your Lamba function
   * It's implement by lambda.Function, ex: NodejsFunction / PythonFunction or CustomFunction
   */
  readonly func?: lambda.Function;
  /**
     * The source code of your Lambda function.
     *
     * You can point to a file in an
     * Amazon Simple Storage Service (Amazon S3) bucket or specify your source
     * code as inline text.
     *
     * @stability stable
     *
     * @default Code.fromAsset(path.join(__dirname, '../lambda/function'))
  */
  readonly code?: lambda.AssetCode;
  /**
     * The runtime environment for the Lambda function that you are uploading.
     *
     * For valid values, see the Runtime property in the AWS Lambda Developer
     * Guide.
     *
     * Use `Runtime.FROM_IMAGE` when when defining a function from a Docker image.
     *
     * @stability stable
     *
     * @default Runtime.PYTHON_3_8
  */
  readonly runtime?: lambda.Runtime;
  /**
     * The name of the method within your code that Lambda calls to execute your function.
     *
     * The format includes the file name. It can also include
     * namespaces and other qualifiers, depending on the runtime.
     * For more information, see https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-features.html#gettingstarted-features-programmingmodel.
     *
     * Use `Handler.FROM_IMAGE` when defining a function from a Docker image.
     *
     * NOTE: If you specify your source code as inline text by specifying the
     * ZipFile property within the Code property, specify index.function_name as
     * the handler.
     *
     * @stability stable
     *
     * @default index.lambda_handler
  */
  readonly handler?: string;
  /**
     * The function execution time (in seconds) after which Lambda terminates the function.
     *
     * Because the execution time affects cost, set this value
     * based on the function's expected execution time.
     *
     * @default Duration.seconds(5)
     * @stability stable
  */
  readonly timeout?: cdk.Duration;
  /**
     * The type of event in response to which should the function be invoked.
     *
     * @stability stable
     *
     * @default LambdaEdgeEventType.ORIGIN_RESPONSE
  */
  readonly eventType?: cf.LambdaEdgeEventType;
  /**
   * The solution identifier
   *
   * @default - no identifier
   */
  readonly solutionId?: string;
  /**
   * The template description
   *
   * @default ''
   */
  readonly templateDescription?: string;
}
/**
 * Custom extension sample
 */
export class Custom extends cdk.NestedStack implements IExtensions {
  readonly functionArn: string;
  readonly functionVersion: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  readonly props: CustomProps;
  constructor(scope: cdk.Construct, id: string, props: CustomProps) {
    super(scope, id, props);

    this.props = props;

    const func = props?.func ?? new lambda.Function(this, 'CustomFunc', {
      code: props?.code ?? lambda.Code.fromAsset(path.join(__dirname, '../lambda/function')),
      runtime: props?.runtime ?? lambda.Runtime.PYTHON_3_8,
      handler: props?.handler ?? 'index.lambda_handler',
      timeout: props?.timeout ?? cdk.Duration.seconds(5),
    });
    this.functionArn = func.functionArn;
    this.functionVersion = func.currentVersion;
    this.eventType = props?.eventType ?? cf.LambdaEdgeEventType.ORIGIN_RESPONSE;
    this._addDescription();
    this._outputSolutionId();
  }
  private _addDescription() {
    this.templateOptions.description = `(${this.props.solutionId}) ${this.props.templateDescription}`;
  }
  private _outputSolutionId() {
    if (this.props.solutionId) {
      new cdk.CfnOutput(this, 'SolutionId', {
        value: this.props.solutionId,
        description: 'Solution ID',
      });
    }
  }
}

/**
 * Generate a lambda function version from the given function ARN
 * @param scope
 * @param id
 * @param functionArn The lambda function ARN
 * @returns lambda.Version
 */
function bumpFunctionVersion(scope: cdk.Construct, id: string, functionArn: string): lambda.Version {
  return new lambda.Version(scope, `LambdaVersion${id}`, {
    lambda: lambda.Function.fromFunctionArn(scope, `FuncArn${id}`, functionArn),
  });
}

/**
 * Default Directory Indexes in Amazon S3-backed Amazon CloudFront Origins
 *
 *  use case - see https://aws.amazon.com/tw/blogs/compute/implementing-default-directory-indexes-in-amazon-s3-backed-amazon-cloudfront-origins-using-lambdaedge/
 */
export class DefaultDirIndex extends Custom {
  readonly lambdaFunction: lambda.Version;
  constructor(scope: cdk.Construct, id: string) {

    super(scope, id, {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.AssetCode.fromAsset(`${EXTENSION_ASSETS_PATH}/cf-default-dir-index`),
      eventType: cf.LambdaEdgeEventType.ORIGIN_REQUEST,
      solutionId: 'SO8134',
      templateDescription: 'Cloudfront extension with AWS CDK - Default Directory Index for Amazon S3 Origin.',
    });
    this.lambdaFunction = this.functionVersion;
  }
};

/**
 * Simple content generation
 * @see https://github.com/awslabs/aws-cloudfront-extensions/tree/main/edge/nodejs/simple-lambda-edge
 */
export class SimpleLambdaEdge extends Custom {
  constructor(scope: cdk.Construct, id: string) {
    const func = new NodejsFunction(scope, 'SimpleLambdaEdgeFunc', {
      entry: `${EXTENSION_ASSETS_PATH}/simple-lambda-edge/index.ts`,
      // L@E does not support NODE14 so use NODE12 instead.
      runtime: lambda.Runtime.NODEJS_12_X,
    });
    super(scope, id, {
      func,
      eventType: cf.LambdaEdgeEventType.VIEWER_REQUEST,
      solutionId: '',
      templateDescription: 'Cloudfront extension with AWS CDK - Simple Lambda Edge.',
    });
  }
};
