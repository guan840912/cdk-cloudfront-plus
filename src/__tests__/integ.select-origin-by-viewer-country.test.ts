import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as origins from '@aws-cdk/aws-cloudfront-origins';
import * as cdk from '@aws-cdk/core';
import * as extensions from '../extensions';


test('minimal usage', () => {
  // GIVEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'demo-stack');

  // WHEN
  // create the cloudfront distribution with extension(s)
  const ext = new extensions.SelectOriginByViwerCountry(stack, 'SelectOriginByCountryCode', {
    countryTable: {
      US: 'mock-api.com',
      CN: 'mock-api.com.cn',
    },
  });

  const policy = new cf.OriginRequestPolicy(stack, 'OrigReqPolicy', {
    headerBehavior: cf.OriginRequestHeaderBehavior.allowList(
      'cloudfront-viewer-country',
    ),
  });

  // create the cloudfront distribution with extension(s)
  const dist = new cf.Distribution(stack, 'dist', {
    defaultBehavior: {
      origin: new origins.HttpOrigin('aws.amazon.com'),
      edgeLambdas: [ext],
      originRequestPolicy: {
        originRequestPolicyId: policy.originRequestPolicyId,
      },
    },
  });

  new cdk.CfnOutput(stack, 'distributionDomainName', {
    value: dist.distributionDomainName,
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      DefaultCacheBehavior: {
        LambdaFunctionAssociations: [
          {
            EventType: 'origin-request',
            LambdaFunctionARN: {
              Ref: 'SelectOriginViewerCountryFuncCurrentVersion32A1D59B0e2ad5c9ee3b77e9947b080cb3c186c1',
            },
          },
        ],
        ViewerProtocolPolicy: 'allow-all',
      },
    },
  });
});
