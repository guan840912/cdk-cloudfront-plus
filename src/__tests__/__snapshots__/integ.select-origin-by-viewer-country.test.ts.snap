// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Outputs": Object {
    "distributionDomainName": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "dist3D2177D6",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecArtifactHash487F0A01": Object {
      "Description": "Artifact hash for asset \\"1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eec\\"",
      "Type": "String",
    },
    "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecS3Bucket25AE18AD": Object {
      "Description": "S3 bucket for asset \\"1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eec\\"",
      "Type": "String",
    },
    "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecS3VersionKey7381E876": Object {
      "Description": "S3 key for asset version \\"1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eec\\"",
      "Type": "String",
    },
    "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4ArtifactHashD6DEF445": Object {
      "Description": "Artifact hash for asset \\"ced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4\\"",
      "Type": "String",
    },
    "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4S3Bucket09F988EA": Object {
      "Description": "S3 bucket for asset \\"ced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4\\"",
      "Type": "String",
    },
    "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4S3VersionKeyD7D49982": Object {
      "Description": "S3 key for asset version \\"ced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "OrigReqPolicyBF71F31A": Object {
      "Properties": Object {
        "OriginRequestPolicyConfig": Object {
          "CookiesConfig": Object {
            "CookieBehavior": "none",
          },
          "HeadersConfig": Object {
            "HeaderBehavior": "whitelist",
            "Headers": Array [
              "cloudfront-viewer-country",
            ],
          },
          "Name": "demostackOrigReqPolicyE18619B9",
          "QueryStringsConfig": Object {
            "QueryStringBehavior": "none",
          },
        },
      },
      "Type": "AWS::CloudFront::OriginRequestPolicy",
    },
    "SelectOriginByCountryCodeNestedStackSelectOriginByCountryCodeNestedStackResource2947FB41": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4S3Bucket09F988EA",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4S3VersionKeyD7D49982",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersced7ef29fc10ca9ff8d83d3a75dc528b0b5e4003835d12d44aefc3a0f7a8d7e4S3VersionKeyD7D49982",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "SelectOriginViewerCountryFuncA5BF3003": Object {
      "DependsOn": Array [
        "SelectOriginViewerCountryFuncServiceRoleBA81B955",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecS3Bucket25AE18AD",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecS3VersionKey7381E876",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters1ed38f40e2e8b97637df0b181112986b486c2b8b315a9c09441599b0ae447eecS3VersionKey7381E876",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "SelectOriginViewerCountryFuncServiceRoleBA81B955",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "SelectOriginViewerCountryFuncCurrentVersion32A1D59B0e2ad5c9ee3b77e9947b080cb3c186c1": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "SelectOriginViewerCountryFuncA5BF3003",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "SelectOriginViewerCountryFuncServiceRoleBA81B955": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "dist3D2177D6": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-request",
                "LambdaFunctionARN": Object {
                  "Ref": "SelectOriginViewerCountryFuncCurrentVersion32A1D59B0e2ad5c9ee3b77e9947b080cb3c186c1",
                },
              },
            ],
            "OriginRequestPolicyId": Object {
              "Ref": "OrigReqPolicyBF71F31A",
            },
            "TargetOriginId": "demostackdistOrigin17D988893",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "CustomOriginConfig": Object {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": "aws.amazon.com",
              "Id": "demostackdistOrigin17D988893",
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;
